File: apps/api/src/schemas/jobs.py
from datetime import datetime
from uuid import UUID

from pydantic import BaseModel, ConfigDict, Field

from src.enums import JobStatus, JobStepStatus, JobType


class JobCreateRequest(BaseModel):
    job_type: JobType
    workspace_id: UUID | None = None
    source_video_id: UUID | None = None
    crawl_session_id: UUID | None = None
    render_output_id: UUID | None = None
    reference_type: str | None = None
    reference_id: UUID | None = None
    idempotency_key: str | None = None
    priority: int = 0
    max_attempts: int = Field(default=3, ge=1, le=20)
    payload_json: dict | None = None
    context_json: dict | None = None


class JobStepResponse(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: UUID
    step_key: str
    step_name: str
    step_order: int
    status: JobStepStatus
    progress_percent: int
    attempts: int
    started_at: datetime | None
    finished_at: datetime | None
    error_code: str | None
    error_message: str | None
    output_json: dict | None


class JobResponse(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: UUID
    workspace_id: UUID
    job_type: JobType
    status: JobStatus
    source_video_id: UUID | None
    crawl_session_id: UUID | None
    render_output_id: UUID | None
    reference_type: str | None
    reference_id: UUID | None
    current_step_key: str | None
    current_step_index: int
    progress_percent: int
    total_steps: int
    completed_steps: int
    failed_steps: int
    priority: int
    attempts: int
    max_attempts: int
    retryable: bool
    locked_by: str | None = None
    locked_at: datetime | None = None
    started_at: datetime | None
    finished_at: datetime | None
    error_code: str | None
    error_message: str | None
    created_at: datetime
    updated_at: datetime
    steps: list[JobStepResponse] = []


class JobListResponse(BaseModel):
    jobs: list[JobResponse]
    total_count: int
    limit: int
    offset: int

File: apps/web/src/types/jobs.ts
export type JobStatus =
  | "QUEUED"
  | "RUNNING"
  | "WAITING_FOR_REVIEW"
  | "RETRYABLE"
  | "FAILED"
  | "CANCELLED"
  | "COMPLETED";

export type JobType =
  | "CRAWL_PROFILE"
  | "SCORE_CANDIDATES"
  | "DOWNLOAD_VIDEO"
  | "ANALYZE_AUDIO"
  | "ANALYZE_OCR"
  | "BUILD_TRANSLATION_DRAFT"
  | "SYNTHESIZE_TTS"
  | "RENDER_PREVIEW"
  | "RENDER_FINAL"
  | "PUBLISH_CONTENT"
  | "REFRESH_PUBLISH_STATUS"
  | "RECONCILE_PUBLISH_ATTEMPT"
  | "COLLECT_PUBLICATION_METRICS"
  | "CLASSIFY_CONTENT"
  | string;

export type JobStep = {
  id: string;
  step_key: string;
  step_name: string;
  step_order: number;
  status: string;
  progress_percent: number;
  attempts: number;
  started_at: string | null;
  finished_at: string | null;
  error_code: string | null;
  error_message: string | null;
  output_json: Record<string, unknown> | null;
};

export type Job = {
  id: string;
  workspace_id: string;
  job_type: JobType;
  workflow_action?: string | null;
  status: JobStatus;
  source_video_id: string | null;
  crawl_session_id: string | null;
  render_output_id: string | null;
  reference_type: string | null;
  reference_id: string | null;
  current_step_key: string | null;
  current_step_index: number;
  progress_percent: number;
  total_steps: number;
  completed_steps: number;
  failed_steps: number;
  priority: number;
  attempts: number;
  max_attempts: number;
  retryable: boolean;
  locked_by?: string | null;
  locked_at?: string | null;
  started_at: string | null;
  finished_at: string | null;
  error_code: string | null;
  error_message: string | null;
  created_at: string;
  updated_at: string;
  steps: JobStep[];
};

export type JobListResponse = {
  jobs: Job[];
  total_count: number;
  limit: number;
  offset: number;
};
